New loader: Load PNG/TIF/JPG microscopy files as numpy array

mariehbourget opened the pull request:
#### GitHub
- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents
- [x] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [x] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR enables the loading of microsocpy PNG files as numpy arrays for training, as well as PNG/TIF/JPG files for inference.
I added 5 methods to the `SegmentationPair` class to replace functions from nibabel library specific to nifti files.

Nibabel functions `nib.load`, `nib.as_closest_canonical`, `handle.header.get_data_shape`, `handle.header.get_zooms`, `handle.get_fdata` are replaced respectively by `read_file`, `apply_canonical`, `get_shape`, `get_voxel_size` and `get_data` throughout the class.

These new methods include a "per file format" behavior based on file extension. For nifti files, the nibabel functions are used as they were before. For PNG/TIF/JPG files, a corresponding behavior is applied. Adding a new modalities of file formats in the future would require to update those 5 functions.

I chose this "minimally invasive" approach because of the complexity of the SegmentationPair class where each nibabel functions is called several times, which I think would require a major refactoring to avoid.
@Drulex, @joshuacwnewton, @ahill187, @gisellemartel, @dyt811 Let us know if you have advice for us on how to approach this refactoring. Thanks!

## To test the PR:
- Any job with nifti files should work as usual
- To test with PNG, use `config_new_loader.json` and the `ivadomed/data_example_microscopy_sem` dataset.

## Linked issues
Fixes #733

mariehbourget committed:
Add extension attribute to SegmentationPair

mariehbourget committed:
Add read_file method to SegmentationPair

mariehbourget committed:
Add apply_canonical method to SegmentationPair

mariehbourget committed:
Add get_shape method to SegmentationPair

mariehbourget committed:
Add get_voxel_size method to SegmentationPair

mariehbourget committed:
Add get_data method to SegmentationPair

mariehbourget committed:
Minor correction

mariehbourget committed:
Update config_new_loader.json

charleygros commented on ivadomed/loader/loader.py:
```diff
@@ -197,12 +199,13 @@ def __init__(self, input_filenames, gt_filenames, metadata=None, slice_axis=2, c
         self.slice_axis = slice_axis
         self.soft_gt = soft_gt
         self.prepro_transforms = prepro_transforms
+        self.extension = input_filenames[0].split('.', 1)[1]
```
Simply curious: why not using this info from the config file?

joshuacwnewton replied:
> Simply curious: why not using this info from the config file?

Alternatively, we could just not store the extension at all. 

* `read_file` could parse the `filename` string.
* The other functions could check the type of the `data` variable. (`Nifti1Image` or `ndarray`)

mariehbourget replied:
This is a very good question!
There are 3 places to fetch the extension:

1- From the config file:
As implemented right now, the extension param accepts multiple extensions as a list of string and not just single one. With multiple formats, we won't be able to tell which file has which extension. Also, the param is optional. If it is empty or not present, all files are indexed.

2 - From the metadata argument: 
That seems to be the best solution. It comes from the BIDSDataframe. However, it would not work for `segment_volume` in `inference.py`. This is use by `sct_deepseg` (if I'm not mistaken) and bypass the BIDS part of the loader entirely.

3 - From the filename:
That is what I implemented at the moment. I realize now, that it may not always work as intended with `segment_volume` if there are other dots in the filenames.

There may be a way to recognize the file type by its content without relying on the filename?

jcohenadad replied:
@mariehbourget to follow-up on our meeting today, this is how we check file type in SCT, with additional checks based on the header (eg: we had bad surprises in the past, where files were NIfTI, but not images):

https://github.com/neuropoly/spinalcordtoolbox/blob/af2d4de531f66cd062af6fefa87c3f84185875ff/spinalcordtoolbox/scripts/sct_apply_transfo.py#L164-L169

mariehbourget replied:
Fixed in 66b2e90

charleygros reviewed (commented):
Looks all good to me! Now testing the code on my pipelines, keep you posted

joshuacwnewton commented on ivadomed/loader/loader.py:
```diff
@@ -424,6 +424,83 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+            data ('nibabel.nifti1.Nifti1Image' object or 'ndarray'):
+                for nifti or png file respectively.
+        """
+        if "nii" in self.extension:
+            # Returns 'nibabel.nifti1.Nifti1Image' object
+            return nib.as_closest_canonical(data)
+        if self.extension == "png":
+            # Returns data as is in numpy array
+            return data
```
Here, we have two `if` statements. What about the `if .. elif .. else raise` pattern like was done in `read_file`?

joshuacwnewton replied:
Also, two `if` statements are used. What about the `if .. elif .. else raise` pattern like was done in `read_file`?

joshuacwnewton commented on ivadomed/loader/loader.py:
```diff
@@ -424,6 +424,83 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+            data ('nibabel.nifti1.Nifti1Image' object or 'ndarray'):
+                for nifti or png file respectively.
+        """
+        if "nii" in self.extension:
+            # Returns 'nibabel.nifti1.Nifti1Image' object
+            return nib.as_closest_canonical(data)
+        if self.extension == "png":
+            # Returns data as is in numpy array
+            return data
```
Instead of checking `self.extension`, it might be more reliable to check the type of the `data` variable, since the behavior is specific to `Nifti1Image` and `ndarray` types.

(This also applies to other functions here.)

joshuacwnewton replied:
Also, two `if` statements are used. What about the `if .. elif .. else raise` pattern like was done in `read_file`?

mariehbourget replied:
I kept the condition on the file extension for now. Because we use BIDS convention, the accepted file formats are limited. For microscopy, we will also want a loader for OMETIFF (not implemented yet), and the behavior with ndarray may differ in some cases (ex: voxel size could be extract from file metadata instead of bids metadata).

I corrected the `if/elif` statements but I did not include an `else` in the functions because `read_file` is the first to be called with the same `if/elif` conditions and would raise the error. Is it bad practice? It would be necessary to add an `else` if we change de `if/elif` condition though.

joshuacwnewton replied:
> For microscopy, we will also want a loader for OMETIFF (not implemented yet), and the behavior with ndarray may differ in some cases (ex: voxel size could be extract from file metadata instead of bids metadata).

Thank you for the clarification. Apologies for making that suggestion without having the full context. :sheep: 

> I corrected the if/elif statements (in d9124b2) but I did not include an else in the functions because read_file is the first to be called with the same if/elif conditions and would raise the error. Is it bad practice?

Good point. I was thinking "what if `read_file` isn't called first?" but I guess there's not much of a risk of that happening, is there...

Still, perhaps we could raise that error in `__init__` when `self.extension` is first set. That way we guarantee the error is thrown before anything else happens (when `loader` is first created).

mariehbourget replied:
> Still, perhaps we could raise that error in `__init__` when `self.extension` is first set. That way we guarantee the error is thrown before anything else happens (i.e. when `loader` is first created).

Fixed in 66b2e90

joshuacwnewton commented on ivadomed/loader/loader.py:
```diff
@@ -424,6 +424,83 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+        Returns:
+            ndarray: File data.
+        """
+        if "nii" in self.extension:
+            # Load data from file as numpy array
+            return data.get_fdata(cache_mode, dtype=np.float32)
+        if self.extension == "png":
+            # Returns data as is in numpy array
+            return data
```
I notice we cast to `float32` in the `Nifti1Image` case, but not in the `ndarray` case. Why is this?

joshuacwnewton commented on ivadomed/loader/loader.py:
```diff
@@ -424,6 +424,83 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+        if "nii" in self.extension:
+            # Read zooms metadata from nifti file header
+            return data.header.get_zooms()
+        if self.extension == "png":
+            # Voxel size for PNG is extracted from PixelSize metadata (from BIDS JSON sidecar)
+            # PixelSize definition in example dataset is a scalar in micrometers (BIDS BEP031 v 0.0.2)
+            # PixelSize definition will change, 2D (XY) or 3D (XYZ) array in micrometers (BIDS BEP031 v 0.0.3)
+            # Behavior will have to be ajusted here to follow BEP development
+            pixel_size_in_mm = self.metadata[0]['PixelSize'] / 1000
+            return (pixel_size_in_mm, pixel_size_in_mm, 0)
```
This worries me a little, because we rely on `PixelSize` to be defined the JSON sidecar. Is there a possibility that this info could ever be missing? 

If so, then perhaps we could provide a more informative error message.

joshuacwnewton replied:
Side question: Does the BIDS JSON sidecar have a field for PixelSize units (e.g. `um` for micrometers)? That would make our life a lot easier, hehe.

mariehbourget replied:
As it is defined now, the `PixelSize` in BIDS JSON sidecar is always in `micrometers`, so I just converted to `mm`. However,  microscopy is not integrated in the BIDS specification yet. We are working on an BIDS extension proposal (BEP) in parallel, so microscopy data/metadata may change.

I updated the function to add an error message and support PixelSize as a float or as a list.

mariehbourget replied:
As defined in BIDS right now 'PixelSize' is always in `micrometer `so I just converted to `mm`. However, microscopy is not integrated in the BIDS specification yet. We are working in parallel on a BIDS extension proposal (BEP) for it and microscopy data/metadata may change.

I added an error message if the field is not present and support for `PixelSize` as a float or as a list.

joshuacwnewton replied:
Thank you for the clarification. :slightly_smiling_face:

joshuacwnewton reviewed (commented):
> @Drulex, @joshuacwnewton, @ahill187, @gisellemartel, @dyt811 Let us know if you have advice for us on how to approach this refactoring. Thanks!

One thing to keep in mind with the "minimally invasive" approach is that it adds ambiguity to the existing variables. For example:

```python
for input_file in self.input_filenames:
    input_img = self.read_file(input_file)  # Q: What is `input_img`? ndarray or a nibabel.Nifti1Image? 
    self.input_handle.append(input_img)
    if len(input_img.shape) > 3:  # Q: Is `input_img` guaranteed to have `shape`? Hard to say by looking at it.
                                  # ANS: They both have `shape`. But what about other attributes/methods?
        raise RuntimeError("4-dimensional volumes not supported.")   
```

This might be an acceptable cost considering the refactoring effort this saves, though.

At the very least, we should make sure to update the docstrings and etc., too. For example:

https://github.com/ivadomed/ivadomed/blob/d578aead456c4c572b909ae5ae650c6c586afc55/ivadomed/loader/loader.py#L187-L188

Not just NifTi images anymore. :sweat_smile:

mariehbourget committed:
Update docstrings

mariehbourget committed:
Update slice_axis loader param in doc

mariehbourget committed:
Correction if/elif statement in new methods

mariehbourget committed:
Update error when file type is not supported

mariehbourget committed:
Update get_voxel_size function for png

mariehbourget commented:
> At the very least, we should make sure to update the docstrings and etc.,

Doctrings are updated in 8693214 along with documentation in 3a72f09.

Thank you @joshuacwnewton for your help!

gisellemartel commented on ivadomed/loader/loader.py:
```diff
@@ -424,14 +425,112 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+            # TODO: Review beahvior when implementing inference with PNG files (no BIDS metadata)
+            if 'PixelSize' in self.metadata[0]:
+                ps_in_um = self.metadata[0]['PixelSize']
+                if isinstance(ps_in_um, list) and (len(ps_in_um) == 2 or len(ps_in_um) == 3):
```
For readability I would suggest to store numerical values in a variable. It would help the maintainer understand why we are checking for in this if statement

mariehbourget replied:
Fixed in dfcdee2

gisellemartel commented on ivadomed/loader/loader.py:
```diff
@@ -424,14 +425,112 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+                else:
+                    raise RuntimeError("'PixelSize' metadata type is not supported. Format must be 2D [X, Y] list, "
+                                       "3D [X, Y, Z] list or float.")
+                ps_in_mm = tuple(ps_in_um / 1000)
```
Same concern as comment above

mariehbourget replied:
Fixed in dfcdee2

gisellemartel commented on ivadomed/loader/loader.py:
```diff
@@ -424,14 +425,112 @@ def get_pair_slice(self, slice_index, gt_type="segmentation"):
+            # PixelSize definition in example dataset is a scalar in micrometers (BIDS BEP031 v 0.0.2)
+            # PixelSize definition may change for 2D [X, Y] and 3D [X, Y, Z] arrays in micrometers (BIDS BEP031 v 0.0.3)
+            # This method supports both behaviors.
+            # TODO: Update behavior to follow BEP microscopy development
+            # TODO: Review beahvior when implementing inference with PNG files (no BIDS metadata)
```
Are these TODOs documented in an issue? If so I would suggest maybe mentioning the issue number here

mariehbourget replied:
Fixed in 253566b

mariehbourget committed:
Merge branch 'master' into mhb/733-load-microscopy-png

mariehbourget committed:
Merge branch 'master' into mhb/733-load-microscopy-png

mariehbourget committed:
Find file extension based on end of filename

mariehbourget committed:
Update if-else statements based on file extensions

mariehbourget committed:
Update docstrings

mariehbourget committed:
Add issue numbers in TODO comments

mariehbourget committed:
Store numerical values in variables

charleygros commented:
> Now testing the code on my pipelines, keep you posted

As said during last meeting: It worked well (just forgot to message you, oopsie)

Let me know if you would like me to re-run tests after your recent changes. Cheers

mariehbourget committed:
Fix loading "as_gray" for tiff files

mariehbourget committed:
Correction file extensions list

mariehbourget committed:
test_loader.py: Add test_load_dataset_2d_png

mariehbourget committed:
Correction to accept file extensions in upper case

mariehbourget commented:
> Let me know if you would like me to re-run tests after your recent changes. Cheers

@charleygros, the PR is ready for re-review.
I've made corrections based on previous comments and sorted out the "extensions" issue.
I also added a test for the loading of 2d PNG data.

Note: the loader now accept PNG, TIF and JPG 2d files for microscopy because we will train with BIDS compliant datasets in PNG but we want to be able to run segmentation from ADS with other formats as well.

Any job with nifti files should work as usual.
For PNG data, only "training" will work at the moment. This is because "testing" and "segment" functions rely heavily on nifti files.  I added comments and issue numbers in the code about that and I will make the necessary changes in future PRs.

charleygros commented:
> @charleygros, the PR is ready for re-review.

I tested on few scenarii of nifiti trainings and... all as expected 🥇 !!

mariehbourget committed:
Merge branch 'master' into mhb/733-load-microscopy-png
